File: LRU_CACHE/lru_cache.js
class LRUCache {
  constructor(capacity) {
    this.capacity = capacity;
    this.cache = {};
    this.head = { key: 'head', next: null, prev: null };
    this.tail = { key: 'tail', next: null, prev: null };
    this.head.next = this.tail;
    this.tail.prev = this.head;
  }

  get(key) {
    if (this.cache[key]) {
      this.moveToHead(this.cache[key]);
      return this.cache[key].value;
    } else {
      return -1; // Key not found
    }
  }

  put(key, value) {
    if (this.cache[key]) {
      this.cache[key].value = value;
      this.moveToHead(this.cache[key]);
    } else {
      if (Object.keys(this.cache).length >= this.capacity) {
        this.removeTail();
      }
      const newNode = { key, value };
      this.cache[key] = newNode;
      this.addToHead(newNode);
    }
  }

  moveToHead(node) {
    this.removeNode(node);
    this.addToHead(node);
  }

  removeNode(node) {
    const prevNode = node.prev;
    const nextNode = node.next;
    prevNode.next = nextNode;
    nextNode.prev = prevNode;
  }

  addToHead(node) {
    const oldHeadNext = this.head.next;
    this.head.next = node;
    node.prev = this.head;
    node.next = oldHeadNext;
    oldHeadNext.prev = node;
  }

  removeTail() {
    const tailNode = this.tail.prev;
    this.removeNode(tailNode);
    delete this.cache[tailNode.key];
  }
}

const cache = new LRUCache(3); // Create an LRU cache with a capacity of 3

cache.put('key1', 'value1');
cache.put('key2', 'value2');
cache.put('key3', 'value3');

console.log(cache.get('key1')); // Output: "value1"
console.log(cache.get('key2')); // Output: "value2"
console.log(cache.get('key3')); // Output: "value3"

cache.put('key4', 'value4'); // This will remove the least recently used item ("key1")

console.log(cache.get('key1')); // Output: -1 (Key not found)

File: LRU_CACHE/lru_cache.py
class LRUCache:
    def __init__(self, capacity) -> None:
        self.capacity = capacity
        self.cache = {}
        self.head = { 'key': 'head', 'next': None, 'prev': None }
        self.tail = { 'key': 'tail', 'next': None, 'prev': None }
        self.head['next'] = self.tail
        self.tail['prev'] = self.head
    

    def get(self, key):
        if (self.cache.get(key) != None):
            self.moveToHead(self.cache.get(key))
            return self.cache[key]['value']
        else:
            return -1
    

    def put(self, key, value):
        if self.cache.get(key) != None:
            self.cache.get(key).value = value
            self.moveToHead(self.cache[key])
        else:
            if (len(self.cache) >= self.capacity):
                self.removeTail()
            newNode = { 'key': key, 'value': value }
            self.cache[key] = newNode
            self.addToHead(newNode)

    def moveToHead(self, node):
        self.removeNode(node)
        self.addToHead(node)
    
    def addToHead(self, node):
        oldHeadNext = self.head.get('next')
        self.head['next'] = node
        node['prev'] = self.head
        node['next'] = oldHeadNext
        oldHeadNext['prev'] = node
    
    def removeNode(self, node):
        prevNode = node['prev']
        nextNode = node['next']
        prevNode['next'] = nextNode
        nextNode['prev'] = prevNode
    
    def removeTail(self):
        tailNode = self.tail.get('prev')
        self.removeNode(tailNode)
        self.cache.pop(tailNode.get('key'))




# Example usage
lru_cache = LRUCache(3)
lru_cache.put(1, 1)
lru_cache.put(2, 2)
lru_cache.put(3, 3)
print(lru_cache.get(2))  # Output: 2
lru_cache.put(4, 4)
print(lru_cache.get(1))  # Output: -1 (not found)
print(lru_cache.get(3)) 